lib/ash/typed_struct.ex

```diff
@@ -181,11 +187,16 @@ defmodule Ash.TypedStruct do
                constraints: unquote(Macro.escape(map_constraints))
 
              @doc "Create a new #{__MODULE__}, raising any error"
+             def new!(unquote(map_required_fields_match) = fields) do
+               fields
+               |> new()
+               |> Ash.Helpers.unwrap_or_raise!()
+             end
+
              def new!(fields) do
-               case new(fields) do
-                 {:ok, value} -> value
-                 {:error, error} -> raise Ash.Error.to_error_class(error)
-               end
+               fields
+               |> new()
+               |> Ash.Helpers.unwrap_or_raise!()
```

Comment:
I think if we want dialyzer or type system help, we'd need something like

`{:error, error} = new(fields)` and `raise Ash.Error.to_error_class(fields)` so that this fall-through *always* raises.